diff.go

```diff
@@ -0,0 +1,146 @@
+// be send to Ralph. Other fields provide convenience shourtcuts for some
+// functions/methods (e.g. Component field holds a reference to the original
+// object, which frees us from unmarshaling contents of Data field).
+type DiffComponent struct {
+	ID        int       // ID of the object held in Component field
+	Name      string    // name of the component (e.g. Ethernet, Memory)
+	Data      []byte    // JSON-ed Component field
+	Component Component // reference to the original object
+}
+
+// NewDiffComponent creates a DiffComponent based on a given component. Since
+// component is an interface type, it can hold both object or pointer, but
+// NewDiffComponent handles both of these cases.
+func NewDiffComponent(component Component) (*DiffComponent, error) {
+	var id int
+	var name string
+	var data []byte
+	var err error
+	switch v := component.(type) {
+	case *Ethernet:
+		id = v.ID
+		name = "Ethernet"
+		data, err = json.Marshal(v)
+		if err != nil {
+			return nil, err
+		}
+	case Ethernet:
+		return NewDiffComponent(&v)
+	case *Memory:
+		id = v.ID
+		name = "Memory"
+		data, err = json.Marshal(v)
+		if err != nil {
+			return nil, err
+		}
+	case Memory:
+		return NewDiffComponent(&v)
+	default:
+		return nil, fmt.Errorf("unknown component: %+v", v)
+	}
+	return &DiffComponent{
+		ID:        id,
+		Name:      name,
+		Data:      data,
+		Component: component,
+	}, nil
+}
+
+func (d DiffComponent) String() string {
+	return fmt.Sprintf("DiffComponent{ID: %d, Name: %s, Data: %s, Component: %s}",
+		d.ID, d.Name, string(d.Data), d.Component)
+}
+
+// SendDiffToRalph sends a given Diff to Ralph. If dryRun is set to true, then
+// no changes will be sent to Ralph. If noOutput is set to true, then all the
+// output from this function will be silenced (this is mostly for tests).
+// Returned statusCodes slice is meant only to facilitate tests, so don't be
+// surprised if you see it ignored somewhere in the source code.
+func SendDiffToRalph(client *Client, diff *Diff, dryRun bool, noOutput bool) (statusCodes []int, err error) {
+
+	var send = func(d *DiffComponent, method, endpoint, msg string) (int, error) {
+		var code int
+		var data []byte
+		switch {
+		case method == "DELETE":
+			data = nil
+		default:
+			data = d.Data
+		}
+		if !dryRun {
+			code, err = client.SendToRalph(method, endpoint, data)
+		}
+		if err != nil {
+			return code, err
+		}
+		if !noOutput {
+			fmt.Printf("%s %s successfully.\n", d.Component, msg) // TODO(xor-xor): Use logger instead.
+		}
+		return code, nil
+	}
+
+	for _, d := range diff.Create {
+		endpoint := APIEndpoints[d.Name]
+		code, err := send(d, "POST", endpoint, "created")
+		if err != nil {
+			return statusCodes, err
+		}
+		if code != 0 {
+			statusCodes = append(statusCodes, code)
+		}
+	}
+	for _, d := range diff.Update {
+		endpoint := fmt.Sprintf("%s/%d", APIEndpoints[d.Name], d.ID)
+		code, err := send(d, "PUT", endpoint, "updated")
```

Comment:
@mkurek - I've just checked `ethernets` and `memory` endpoints, and it seems that for Ralph/DjRF, PUT behaves just like PATCH - i.e., it doesn't require all the fields, and omitted fields are not overwritten (i.e., zeroed out).

And as a side note - taking the aforementioned RFC6902 aside, PATCH is kind of troublesome with Go due to its *zero values* (e.g., `0` for `int`, `""` for `string`, `false` for `bool`, `nil` for pointers etc. - there's no such thing as `void` type in C - and that's for good reasons, though). But actually, it seems that there's some quite elegant solution for this problem (see [this article](https://willnorris.com/2014/05/go-rest-apis-and-pointers) [1]), but I need to analyze how easily/hard it would be to implement this in `ralph-cli`.


[1] It also mentions that, for example, GitHub's API doesn't allow PUT at all (!) - when you want to update some resource, you have to send the whole object via PATCH method (!!), which is semantically equal to PUT :wink: And BTW, in DjRF before 3.x version, PUT was used to create objects, just like POST... So yeah, it seems that everyone has its own interpretation of what REST really is... RFC, anyone..? :wink:

(CC: @vi4m)